src/storage/restrict.rs

```diff
@@ -146,152 +158,470 @@ where
+
+    unsafe fn open(self) -> (Self::Mask, Self::Value) {
+        let bitset = self.bitset.borrow();
+        (bitset, (self.data.borrow(), self.entities, bitset))
+    }
+
+    unsafe fn get(value: &Self::Value, id: Index) -> Self::Type {
+        // NOTE: Methods on this type rely on safety requiments of this method.
+        PairedStorageRead {
+            index: id,
+            storage: value.0,
+            entities: value.1,
+            bitset: value.2,
+        }
     }
 }
 
-impl<'rf, 'st, C, S, B, Restrict> PairedStorage<'rf, 'st, C, S, B, Restrict>
+// SAFETY: It is safe to call `get` from multiple threads at once since
+// `T::Storage: Sync`. We construct a `PairedStorageSharedWrite` which can be
+// used to call `UnprotectedStorage::get` which is safe to call concurrently and
+// `SharedGetOnly::get_mut` which is safe to call concurrently since we require
+// `C::Storage: DistinctStorage` here.
+//
+// `open` returns references to corresponding mask and storage values contained
+// in the wrapped `Storage`.
+//
+// Iterating the mask does not repeat indices.
+#[cfg(feature = "parallel")]
+unsafe impl<'rf, C, S> ParJoin for &'rf mut RestrictedStorage<'rf, C, S>
 where
     C: Component,
     S: BorrowMut<C::Storage>,
-    B: Borrow<BitSet>,
+    C::Storage: Sync + SharedGetMutStorage<C> + DistinctStorage,
 {
-    /// Gets the component related to the current entry without checking whether
-    /// the storage has it or not.
-    pub fn get_mut_unchecked(&mut self) -> AccessMutReturn<'_, C> {
-        unsafe { self.storage.borrow_mut().get_mut(self.index) }
+    type Mask = &'rf BitSet;
+    type Type = PairedStorageWriteShared<'rf, C>;
+    type Value = SharedGetOnly<'rf, C, C::Storage>;
+
+    unsafe fn open(self) -> (Self::Mask, Self::Value) {
+        let bitset = &self.bitset;
+        let storage = SharedGetOnly::new(self.data.borrow_mut());
+        (bitset, storage)
     }
+
+    unsafe fn get(value: &Self::Value, id: Index) -> Self::Type {
+        // NOTE: Methods on this type rely on safety requiments of this method.
+        PairedStorageWriteShared {
+            index: id,
+            storage: SharedGetOnly::duplicate(value),
+        }
+    }
+}
+
+/// Pairs a storage with an index, meaning that the index is guaranteed to exist
+/// as long as the `PairedStorage<C>` exists.
+///
+/// Yielded by `lend_join`/`join`/`par_join` on `&storage.restrict()`.
+pub struct PairedStorageRead<'rf, C: Component> {
+    index: Index,
+    storage: &'rf C::Storage,
+    bitset: &'rf BitSet,
+    entities: &'rf Fetch<'rf, EntitiesRes>,
+}
+
+/// Pairs a storage with an index, meaning that the index is guaranteed to
+/// exist.
+///
+/// Yielded by `join`/`par_join` on `&mut storage.restrict_mut()`.
+pub struct PairedStorageWriteShared<'rf, C: Component> {
+    index: Index,
+    storage: SharedGetOnly<'rf, C, C::Storage>,
 }
 
-impl<'rf, 'st, C, S, B, Restrict> PairedStorage<'rf, 'st, C, S, B, Restrict>
```

Comment:
With the additional complexity of adding the `LendJoin` case it was much easier to have several separate paired storage types rather than using generic marker parameters.